scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala

```diff
@@ -363,6 +365,89 @@ class FormatWriter(formatOps: FormatOps) {
       finalResult.result()
     }
   }
+
+  import scala.meta.internal.classifiers.classifier
+
+  @classifier
+  trait ClosingBracket
+  object ClosingBracket {
+    def unapply(token: Token): Boolean =
+      token.is[RightParen] || token.is[RightBracket] || token.is[RightBrace]
+  }
+
+  private def handleTrailingCommasAndWhitespace(
+      formatToken: FormatToken,
+      state: State,
+      sb: StringBuilder,
+      whitespace: String): Unit = {
+
+    import org.scalafmt.config.TrailingCommas
+
+    if (!runner.dialect.allowTrailingCommas) {
+      sb.append(whitespace)
+      return
+    }
+
+    val left = formatToken.left
+    val right = nextNonComment(formatToken).right
+    val isNewline = state.splits.last.modification.isNewline
+
+    initStyle.trailingCommas match {
+      // foo(
+      //   a,
+      //   b
+      // )
+      //
+      // Insert a comma after b
+      case TrailingCommas.always
+          if !left.is[Comma] && !left.is[Comment] &&
+            right.is[ClosingBracket] && isNewline =>
+        sb.append(",")
+        sb.append(whitespace)
+
+      // foo(
+      //   a,
+      //   b // comment
+      // )
+      //
+      // Insert a comma after b (before comment)
+      case TrailingCommas.always
+          if left.is[Comment] && !prev(left).is[Comma] && right.is[ClosingBracket] && isNewline =>
+        sb.insert(sb.length - left.syntax.length - 1, ",")
+        sb.append(whitespace)
```

Comment:
I had to deal with comments in this way, detecting a comment, looking for a comma next to it, going back in the buffer and removing the comma.

Not sure how safe is this.

I had initially tried to incorporate this into the previous case, but in presence of `[ID][COMMENT][NEWLINE][CLOSING_BRACKET]` I couldn't detect the newline after che comment.